File: plc4j/protocols/driver-bases/base/src/test/java/org/apache/plc4x/java/base/model/DefaultPlcConsumerRegistrationTest.java
/*
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 */

package org.apache.plc4x.java.base.model;

import nl.jqno.equalsverifier.EqualsVerifier;
import org.apache.plc4x.java.api.messages.PlcSubscriptionEvent;
import org.apache.plc4x.java.base.messages.PlcSubscriber;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Collection;
import java.util.function.Consumer;

import static org.hamcrest.core.IsEqual.equalTo;
import static org.hamcrest.core.IsNull.notNullValue;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class DefaultPlcConsumerRegistrationTest {

    @Mock
    private PlcSubscriber subscriber;
    @Mock
    private Consumer<PlcSubscriptionEvent> consumer;
    @Mock
    private InternalPlcSubscriptionHandle subscriptionHandle;
    private DefaultPlcConsumerRegistration SUT;

    @BeforeEach
    void setUp() {
        SUT = new DefaultPlcConsumerRegistration(subscriber, consumer, subscriptionHandle);
    }

    @Test
    void getConsumerHash() {
        int consumerHash = SUT.getConsumerHash();
        assertThat(consumerHash, equalTo(consumer.hashCode()));
    }

    @Test
    void getAssociatedHandles() {
        Collection<InternalPlcSubscriptionHandle> associatedHandles = SUT.getAssociatedHandles();
        assertThat(associatedHandles, notNullValue());
        assertThat(associatedHandles.size(), equalTo(1));
        assertThat(associatedHandles.iterator().next(), equalTo(subscriptionHandle));
    }

    @Test
    public void unregister() {
        SUT.unregister();
        verify(subscriber, times(1)).unregister(any());
    }

    @Test
    public void testToString() {
        String s = SUT.toString();
        assertThat(s, notNullValue());
    }

    @Test
    void equals() {
        EqualsVerifier.forClass(DefaultPlcConsumerRegistration.class).verify();
    }

}

File: plc4go/spi/model/DefaultPlcConsumerRegistration_test.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package model

import (
	apiModel "github.com/apache/plc4x/plc4go/pkg/api/model"
	"github.com/apache/plc4x/plc4go/spi"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
	"testing"
)

func TestDefaultPlcConsumerRegistration_GetConsumerId(t *testing.T) {
	type fields struct {
		consumerId    int
		consumer      apiModel.PlcSubscriptionEventConsumer
		plcSubscriber spi.PlcSubscriber
		handles       []apiModel.PlcSubscriptionHandle
	}
	tests := []struct {
		name   string
		fields fields
		want   int
	}{
		{
			name: "get it",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := &DefaultPlcConsumerRegistration{
				consumerId:    tt.fields.consumerId,
				consumer:      tt.fields.consumer,
				plcSubscriber: tt.fields.plcSubscriber,
				handles:       tt.fields.handles,
			}
			assert.Equalf(t, tt.want, d.GetConsumerId(), "GetConsumerId()")
		})
	}
}

func TestDefaultPlcConsumerRegistration_GetSubscriptionHandles(t *testing.T) {
	type fields struct {
		consumerId    int
		consumer      apiModel.PlcSubscriptionEventConsumer
		plcSubscriber spi.PlcSubscriber
		handles       []apiModel.PlcSubscriptionHandle
	}
	tests := []struct {
		name   string
		fields fields
		want   []apiModel.PlcSubscriptionHandle
	}{
		{
			name: "get it",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := &DefaultPlcConsumerRegistration{
				consumerId:    tt.fields.consumerId,
				consumer:      tt.fields.consumer,
				plcSubscriber: tt.fields.plcSubscriber,
				handles:       tt.fields.handles,
			}
			assert.Equalf(t, tt.want, d.GetSubscriptionHandles(), "GetSubscriptionHandles()")
		})
	}
}

func TestDefaultPlcConsumerRegistration_Unregister(t *testing.T) {
	type fields struct {
		consumerId    int
		consumer      apiModel.PlcSubscriptionEventConsumer
		plcSubscriber spi.PlcSubscriber
		handles       []apiModel.PlcSubscriptionHandle
	}
	tests := []struct {
		name   string
		fields fields
	}{
		{
			name: "unregister it",
			fields: fields{
				plcSubscriber: func() spi.PlcSubscriber {
					subscriber := NewMockPlcSubscriber(t)
					subscriber.EXPECT().Unregister(mock.Anything).Return()
					return subscriber
				}(),
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			d := &DefaultPlcConsumerRegistration{
				consumerId:    tt.fields.consumerId,
				consumer:      tt.fields.consumer,
				plcSubscriber: tt.fields.plcSubscriber,
				handles:       tt.fields.handles,
			}
			d.Unregister()
		})
	}
}

func TestNewDefaultPlcConsumerRegistration(t *testing.T) {
	type args struct {
		plcSubscriber spi.PlcSubscriber
		consumer      apiModel.PlcSubscriptionEventConsumer
		handles       []apiModel.PlcSubscriptionHandle
	}
	tests := []struct {
		name string
		args args
		want apiModel.PlcConsumerRegistration
	}{
		{
			name: "create it",
			want: &DefaultPlcConsumerRegistration{},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := NewDefaultPlcConsumerRegistration(tt.args.plcSubscriber, tt.args.consumer, tt.args.handles...).(*DefaultPlcConsumerRegistration)
			tt.want.(*DefaultPlcConsumerRegistration).consumerId = got.consumerId
			assert.Equalf(t, tt.want, got, "NewDefaultPlcConsumerRegistration(%v, func(), %v)", tt.args.plcSubscriber, tt.args.handles)
		})
	}
}
